### PR TITLE
Added recursiveness, improved parsing, added errorhandling for chapters

### PR DESCRIPTION
**Pull request by nikolas-sturm**
Tested on my system. Seems to work fine. Haven't used Python in quite a while.

### CODE REVIEW

**Comment by nikolas-sturm** [comment]
also: chapters can now also be without the 'c' in front

and i added automatic directory structuring by name

**Comment by nikolas-sturm** [comment]
oh and chapters like 001x2 work now

(sry did this all in one swoop this morning)

File: ExtractChaptersFromVolume.py [hunk 1]
```diff
@@ -28,8 +31,9 @@
 
 ext = '.cbz'
 cwd = args.dir
-vol_regex = r'(?:v)[\d)]+'
-ch_regex = r'c([^\s]+)'
+series_regex = r'^.+?(?= [([{]| v[0-9]+)'
```
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: What filenames have you tested this with? None of mine have been able to parse

I assume the series parsing is done to create sub-folders when using `-r`? I'd personally rather go with creating a sub-folder relative to the volume being extracted due to [how complex](https://github.com/Kareadita/Kavita/blob/develop/API/Services/Tasks/Scanner/Parser/Parser.cs) series parsing can get.

I'll add a commit to show what I mean. Lmk what you think of it

File: ExtractChaptersFromVolume.py [hunk 2]
```diff
@@ -41,26 +45,47 @@ def cleanup(extract_path=None):
...
-                volume = re.search(vol_regex, file).group(0).lstrip('v').rstrip(')').lstrip('0').zfill(2)
-                chapter = re.search(ch_regex, file).group(0).lstrip('c').lstrip('0').zfill(3)
+                volume = re.search(vol_regex, file).group(0).lstrip(' (v0').rstrip(') ').zfill(2)
+                try:
+                    chapter = re.search(ch_regex, file).group(0).lstrip(' (c0').rstrip(') ')
+                except AttributeError:
+                    print(f"No chapters found in \"{file}\". Skipping...")
+                    not_found_counter += 1
+                    if not_found_counter >= 3:
+                        print(f"-> There seem to be not chapters in \"{volume_file[1]}\\{volume_file[0]}\". Skipping...")
```
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: "not chapters" -> "no chapters"

File: ExtractChaptersFromVolume.py [hunk 3]
```diff
@@ -83,14 +108,22 @@ def cleanup(extract_path=None):
...
 
                 output_folder = (ext_dest / chapter_folder).resolve()
-                output_file = f"{(cwd / 'output' / chapter_folder).resolve()}{ext}"
+                output_file = f"{(cwd / 'output' / series_name / chapter_folder).resolve()}{ext}"
+
+                if not os.path.exists(f"{(cwd / 'output' / series_name).resolve()}"): 
+                    os.mkdir(f"{(cwd / 'output' / series_name).resolve()}")
 
-                print(f'Creating: {chapter_folder}{ext}')
+                print(f'Creating: \"{chapter_folder}{ext}\"')
```
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: The `\` escapes aren't needed

File: ExtractChaptersFromVolume.py (hunk 2 above)
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: ~~The \ escapes aren't needed~~ you can just swap the outer double quotes for single quotes

File: ExtractChaptersFromVolume.py [hunk 4]
```diff
@@ -41,26 +45,47 @@ def cleanup(extract_path=None):
...
         for file in os.listdir(ext_dest):
             if not file.endswith(('.xml', '.bak')):  # Ignore ComicInfo and the backup that metadata-manager makes
                 log(file)
-                volume = re.search(vol_regex, file).group(0).lstrip('v').rstrip(')').lstrip('0').zfill(2)
-                chapter = re.search(ch_regex, file).group(0).lstrip('c').lstrip('0').zfill(3)
+                volume = re.search(vol_regex, file).group(0).lstrip(' (v0').rstrip(') ').zfill(2)
+                try:
+                    chapter = re.search(ch_regex, file).group(0).lstrip(' (c0').rstrip(') ')
+                except AttributeError:
+                    print(f"No chapters found in \"{file}\". Skipping...")
```
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: ~~The \ escapes aren't needed~~ you can just swap the outer double quotes for single quotes

File: ExtractChaptersFromVolume.py [hunk 5]
```diff
@@ -41,26 +45,47 @@ def cleanup(extract_path=None):
...
+        series_name = re.search(series_regex, volume_file[0]).group(0)
+        ext_dest = ext_folder / volume_file[0].rsplit('.', 1)[0]
         if not ext_dest.exists():
             Path.mkdir(ext_dest, parents=True, exist_ok=True)
 
-        with zipfile.ZipFile(cwd / volume_file, 'r') as f:
+        with zipfile.ZipFile(cwd / volume_file[1] / volume_file[0], 'r') as f:
             f.extractall(ext_dest)
-            print(f"Processing: {volume_file}")
+            print(f"\nProcessing: \"{volume_file[1]}\\{volume_file[0]}\"")
```
Comment (TheIceCreamTroll) [inline ExtractChaptersFromVolume.py]: ~~The `\` escapes aren't needed~~ you can just swap the outer double quotes for single quotes

**Review by TheIceCreamTroll — CHANGES_REQUESTED** [review]
Love the regex changes! Just need to cleanup the string escapes, and take another look at how we handle sub-folders for recursion

File: ExtractChaptersFromVolume.py (hunk 1 above)
Comment (TheIceCreamTroll) [reply ExtractChaptersFromVolume.py]: Ah, I don't think "Allow Edits From Maintainers" is checked. Could you enable it?

Comment (nikolas-sturm) [reply ExtractChaptersFromVolume.py]: Allow edits by maintainers was checked and is still checked... Don't know why it's not working for you.

I only used the series name parsing using files structured like this: "Berserk v40 ....cbz" or "Berserk [....cbz". All my cbz's start in one of these ways, but maybe this isn't suitable for the public script.
Also "series" might be a bit misleading: It creates seperate folders per "season", so "Tokyo Ghoul" and "Tokyo Ghoul - re" are seperate folders.

Comment (TheIceCreamTroll) [reply ExtractChaptersFromVolume.py]: Ah, it was just Git gore on my end. Give my commit a run and lmk what you think

Now that you've explained it, I can see the use case, although I think that the regex too finicky to be the default behavior. "Berserk Vol. 40" and "Berserk Volume 40", for example, are two fairly commonly used cases that fail. If you want to add it as a CLI arg though, to override the default behavior, I wouldn't be against it.

Comment (nikolas-sturm) [reply ExtractChaptersFromVolume.py]: Made the series parsing optional and made the regex way more robust. Lmk if you're happy with this 👍🏻

Comment (TheIceCreamTroll) [reply ExtractChaptersFromVolume.py]: Perfect! I'll go ahead and merge. Just changed a few of the argparse descriptions to be a bit more clear.

Thanks again for opening this PR. Lots of great improvements!

**Review by TheIceCreamTroll — APPROVED** [review]

**Pull request merged**